unumpy/_multimethods.py

```diff
@@ -784,6 +824,69 @@ def broadcast_to(array, shape, subok=False):
     return (array,)
 
 
+def _normalize_axis(ndim, axis, allow_repeats=False):
+    if isinstance(axis, int):
```

Comment:
Can you use [`isinstance(axis, numbers.Number)`](https://docs.python.org/3/library/numbers.html). That will allow custom numeric types like `numpy.int64` to work as axes. Then call `operator.index` in the `axis_pos` comprehension to convert to a standard python `int`. That will also raise an error if it's not a valid index value, e.g. `operator.index(3.2)` will raise where `int(3.2)` wouldn't.

Reply:
[`numbers.Integral`](https://docs.python.org/3/library/numbers.html#numbers.Integral)?

Reply:
Can you explain why `numbers.Integral` is a better choice?

Reply:
The best illustration is often code:
```python
In [2]: issubclass(float, numbers.Number)
Out[2]: True

In [3]: issubclass(complex, numbers.Number)
Out[3]: True
```

Reply:
I actually think `Integral` is not as good here because if you do pass in an floating point number then you will get this error:
```
TypeError: 'float' object is not iterable
```

Whereas with `Number` and `operator.index` you get this error:
```
TypeError: 'float' object cannot be interpreted as an integer
```
Which actually matches the root cause of the failure.

Reply:
Of course you could also raise a custom error and both will be fine.

Reply:
You get the same error with NumPy, which is more-or-less what we'd like:

```
In [1]: import numpy as np

In [2]: np.sum([1], axis=1.3)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-77d3876b85df> in <module>
----> 1 np.sum([1], axis=1.3)

<__array_function__ internals> in sum(*args, **kwargs)

~/miniforge3/envs/uarray/lib/python3.7/site-packages/numpy/core/fromnumeric.py in sum(a, axis, dtype, out, keepdims, initial, where)
   2227
   2228     return _wrapreduction(a, np.add, 'sum', axis, dtype, out, keepdims=keepdims,
-> 2229                           initial=initial, where=where)
   2230
   2231

~/miniforge3/envs/uarray/lib/python3.7/site-packages/numpy/core/fromnumeric.py in _wrapreduction(obj, ufunc, method, axis, dtype, out, **kwargs)
     88                 return reduction(axis=axis, out=out, **passkwargs)
     89
---> 90     return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
     91
     92

TypeError: 'float' object cannot be interpreted as an integer
```

Reply:
I think I understand now.

```python
In [3]: isinstance(1.0, numbers.Integral)                                       
Out[3]: False

In [4]: isinstance(1+0j, numbers.Integral)                                      
Out[4]: False

In [5]: operator.index(1.0)                                                     
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-85c4da9291d5> in <module>
----> 1 operator.index(1.0)

TypeError: 'float' object cannot be interpreted as an integer
```
This safeguards the `operator.index` from failing inside the list comprehension. Maybe also add an `else` clause that raises a `TypeError` for something other than `numbers.Integral`?

Reply:
> You get the same error with NumPy, which is more-or-less what we'd like:

`numbers.Number` seems the right choice for me.